lib/pbench/server/database/models/datasets.py

```diff
@@ -656,6 +457,54 @@ def delete(self):
             raise
 
 
+class OperationName(enum.Enum):
+    """Track and orchestrate the progress of a dataset (tarball) through the
+    various stages of the Pbench server.
+    """
+
+    BACKUP = enum.auto()
+    DELETE = enum.auto()
+    INDEX = enum.auto()
+    REINDEX = enum.auto()
+    TOOLINDEX = enum.auto()
+    UNPACK = enum.auto()
+    UPLOAD = enum.auto()
+
+
+class OperationState(enum.Enum):
+    """Record the status of an Operation as its enabled and retired"""
+
+    READY = enum.auto()
+    WORKING = enum.auto()
+    OK = enum.auto()
+    FAILED = enum.auto()
+
+
+class Operation(Database.Base):
+    """Orchestrate and track the operational flow of datasets through the
+    server.
+
+    This table is managed by the Sync class, but defined here with the parent
+    Dataset class since they're linked.
+
+    Columns:
+        id          Generated unique ID of table row
+        dataset_ref Dataset row ID (foreign key)
+        name        Operation name (OperationName enum)
+        status      Status of operation (OperationStatus enum)
+        message     Message explaining operation status
+    """
+
+    __tablename__ = "dataset_operations"
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    name = Column(Enum(OperationName), index=True)
+    state = Column(Enum(OperationState))
+    message = Column(String(255))
+    dataset_ref = Column(Integer, ForeignKey("datasets.id"))
+    dataset = relationship("Dataset", back_populates="operations")
```

Comment:
`dataset` is a convenient join artifact of SQLAlchemy's imagination, while `dataset_ref` is the actual SQL foreign key. Sometimes, having a collection of related objects on the parent is convenient. I started out without this, but found it convenient at one point. We do this in other places, too.

Reply:
I think because this model follows a one-to-many relationship (one dataset and multiple operations) so we need a ForeignKey column in this table referencing the dataset table. 

We need `relationship` in both tables because we need to perform Joins on both tables. Without the relationship our ORM wouldn't be able to Join with other table. So I think Foreign Key is a SQL operation and relationship is an ORM operation so we need both if that make sense.